### ISSUE TITLE
build: error generating docs in new dev environment

### ISSUE THREAD

**Issue by santind** [ibis-project/ibis#4303]
I'm following the [development environment setup instructions for Conda](https://ibis-project.org/docs/3.0.2/contribute/01_environment/#support-matrix_1) to bootstrap a new environment on an M1 MacBook Pro.  When I attempt to run `mkdocs serve` per the instructions, it reports the error below.  I get the same error under Conda for both Intel and M1 architectures, and it's not completely obvious to me whether this is related to my environment setup or if the docs on master are broken.  Any thoughts on what could be wrong?
```
$ mkdocs serve
INFO     -  Building documentation...
INFO     -  [macros] - Macros arguments: {'module_name': 'main', 'modules': [], 'include_dir': '', 'include_yaml': [],
            'j2_block_start_string': '', 'j2_block_end_string': '', 'j2_variable_start_string': '', 'j2_variable_end_string': '',
            'on_undefined': 'keep', 'on_error_fail': False, 'verbose': False}
INFO     -  [macros] - Extra variables (config file): ['version', 'project_name', 'team', 'social', 'support_levels']
INFO     -  [macros] - Extra filters (module): ['pretty']
INFO     -  Cleaning site directory
INFO     -  Converting notebook (execute=True): /Users/dsantin/Documents/Ibis/ibis/docs/ibis-for-sql-programmers.ipynb
ERROR    -  Error reading page 'ibis-for-sql-programmers.ipynb': format() argument must be a formatter instance, not a class
Traceback (most recent call last):
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/pygments/__init__.py", line 61, in format
    formatter.format(tokens, realoutfile)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/pygments/formatter.py", line 94, in format
    return self.format_unencoded(tokensource, outfile)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/pygments/formatters/html.py", line 981, in format_unencoded
    source = self.wrap(source)
TypeError: wrap() missing 1 required positional argument: 'outfile'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/bin/mkdocs", line 10, in <module>
    sys.exit(cli())
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs/__main__.py", line 181, in serve_command
    serve.serve(dev_addr=dev_addr, livereload=livereload, watch=watch, **kwargs)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs/commands/serve.py", line 63, in serve
    config = builder()
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs/commands/serve.py", line 58, in builder
    build(config, live_server=live_server, dirty=dirty)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs/commands/build.py", line 292, in build
    _populate_page(file.page, config, files, dirty)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs/commands/build.py", line 174, in _populate_page
    page.render(config, files)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs_jupyter/plugin.py", line 94, in new_render
    body = convert.nb2html(
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs_jupyter/nbconvert2.py", line 105, in nb2html
    content, resources = exporter.from_filename(nb_path)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/nbconvert/exporters/exporter.py", line 189, in from_filename
    return self.from_file(f, resources=resources, **kw)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/nbconvert/exporters/exporter.py", line 206, in from_file
    return self.from_notebook_node(
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/nbconvert/exporters/html.py", line 215, in from_notebook_node
    return super().from_notebook_node(nb, resources, **kw)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/nbconvert/exporters/templateexporter.py", line 400, in from_notebook_node
    output = self.template.render(nb=nb_copy, resources=resources)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs_jupyter/templates/mkdocs_html/notebook.html.j2", line 78, in top-level template code
    {%- set vars = {"cell_id": 0 } -%}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/share/jupyter/nbconvert/templates/lab/index.html.j2", line 3, in top-level template code
    {% from 'jupyter_widgets.html.j2' import jupyter_widgets %}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/share/jupyter/nbconvert/templates/lab/base.html.j2", line 260, in top-level template code
    {% set div_id = uuid4() %}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/share/jupyter/nbconvert/templates/base/display_priority.j2", line 1, in top-level template code
    {%- extends 'base/null.j2' -%}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/share/jupyter/nbconvert/templates/base/null.j2", line 26, in top-level template code
    {%- block body -%}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/share/jupyter/nbconvert/templates/base/null.j2", line 29, in block 'body'
    {%- block body_loop -%}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/share/jupyter/nbconvert/templates/base/null.j2", line 31, in block 'body_loop'
    {%- block any_cell scoped -%}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/share/jupyter/nbconvert/templates/base/null.j2", line 87, in block 'any_cell'
    {%- block markdowncell scoped-%} {%- endblock markdowncell -%}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/share/jupyter/nbconvert/templates/lab/base.html.j2", line 101, in block 'markdowncell'
    {{ cell.source  | markdown2html | strip_files_prefix }}
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs_jupyter/nbconvert2.py", line 200, in custom_markdown2html
    return MarkdownWithMath(
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mistune.py", line 1001, in render
    return self.parse(text)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mistune.py", line 1004, in parse
    out = self.output(preprocessing(text))
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mistune.py", line 1053, in output
    out += self.tok()
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mistune.py", line 1063, in tok
    return getattr(self, 'output_%s' % t)()
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mistune.py", line 1085, in output_code
    return self.renderer.block_code(
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/mkdocs_jupyter/nbconvert2.py", line 219, in block_code
    return highlight(code, lexer, formatter)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/pygments/__init__.py", line 82, in highlight
    return format(lex(code, lexer), formatter, outfile)
  File "/Users/dsantin/miniconda3-intel/envs/ibis-dev-intel/lib/python3.9/site-packages/pygments/__init__.py", line 69, in format
    raise TypeError('format() argument must be a formatter instance, '
TypeError: format() argument must be a formatter instance, not a class
```

**Comment by santind** [comment]
I tested by going back to tag 3.0.0 and the environment seems to be okay.  `pytest -m core` runs successfully on 3.0.0 but is broken on HEAD due to missing sqlite DB files. `mkdocs serve` fails the same way for me all the way back to 1e69def.

**Comment by anjakefala** [comment]
I think this is less a problem on the ibis codebase end, and more to do with our recent version pins for the dependencies. With my older environment, `mkdocs serve` builds fine.

A cursory search came upon this: https://github.com/pygments/pygments/issues/2119. `pygments` had an API change.

**Comment by cpcloud** [comment]
Hm, the CI builds the docs successfully in CI and has pygments 2.12 in the environment. I also am able to run `mkdocs serve` without issue.

Both the CI (for the doc build) and my local environment are using the `nix` setup, which may be an important factor here.

I think the issue is not with pygments, but `mkdocs-jupyter`.

In the poetry lockfile, `mkdocs-jupyter` has version `0.21.0` while in the conda lockfile `mkdocs-jupyter` has version `0.20.0`.

It looks like the [conda-forge feedstock](https://github.com/conda-forge/mkdocs-jupyter-feedstock) is out of date.

@santind To unblock yourself, can you try `pip install mkdocs-jupyter==0.21.0` in your conda environment?

In the meantime, I'll look into getting the feedstock updated.

**Comment by cpcloud** [comment]
PR to update `mkdocs-jupyte-feedstock` is [here](https://github.com/conda-forge/mkdocs-jupyter-feedstock/pull/17).

**Comment by cpcloud** [comment]
Alright, I've merged https://github.com/conda-forge/mkdocs-jupyter-feedstock/pull/17. In a couple hours I'll check back to see if the package is available on conda-forge and then run the ibis CI job that regenerates the conda lockfiles to get the latest version.

**Comment by anjakefala** [comment]
Ah, that makes sense! `pygments` had the API change, and that was the version of `mkdocs-jupyter` that updated with it.

`master` builds and serves successfully for me.

**Comment by cpcloud** [comment]
@santind I'm going to close this out for now! If you encounter any more problems, please open another issue. Much appreciated!

**Comment by santind** [comment]
Just wanted to confirm that `master` builds and serves docs for me as well. Thank you @cpcloud and @anjakefala!
